gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py

```diff
@@ -0,0 +1,699 @@
+
+    def update_policy_target_removed(self, context, policy_target):
+        if context.current_profile['service_type'] == pconst.LOADBALANCER:
+            if self._is_service_target(policy_target):
+                return
+            context._plugin_context = self._get_resource_owner_context(
+                context._plugin_context)
+            network_function_map = (
+                self._get_node_instance_network_function_map(
+                    context.plugin_session,
+                    context.current_node['id'],
+                    context.instance['id']))
+
+            if network_function_map:
+                network_function_id = network_function_map.network_function_id
+                self.nfp_notifier.policy_target_removed_notification(
+                    context.plugin_context, network_function_id, policy_target)
+
+    def notify_chain_parameters_updated(self, context):
+        pass  # We are not using the classifier specified in redirect Rule
+
+    def update_node_consumer_ptg_added(self, context, policy_target_group):
+        if context.current_profile['service_type'] == pconst.FIREWALL:
+            context._plugin_context = self._get_resource_owner_context(
+                context._plugin_context)
+            network_function_map = (
+                self._get_node_instance_network_function_map(
+                    context.plugin_session,
+                    context.current_node['id'],
+                    context.instance['id']))
+
+            if network_function_map:
+                network_function_id = network_function_map.network_function_id
+                self.nfp_notifier.consumer_ptg_added_notification(
+                    context.plugin_context,
+                    network_function_id,
+                    policy_target_group)
+
+    def update_node_consumer_ptg_removed(self, context, policy_target_group):
+        if context.current_profile['service_type'] == pconst.FIREWALL:
+            context._plugin_context = self._get_resource_owner_context(
+                context._plugin_context)
+            network_function_map = (
+                self._get_node_instance_network_function_map(
+                    context.plugin_session,
+                    context.current_node['id'],
+                    context.instance['id']))
+
+            if network_function_map:
+                network_function_id = network_function_map.network_function_id
+                self.nfp_notifier.consumer_ptg_removed_notification(
+                    context.plugin_context,
+                    network_function_id,
+                    policy_target_group)
+
+    def _wait_for_network_function_delete_completion(self, context,
+                                                     network_function_id):
+        time_waited = 0
+        network_function = None
+        while time_waited < cfg.CONF.nfp_node_driver.service_delete_timeout:
+            network_function = self.nfp_notifier.get_network_function(
```

Comment:
Would it be better to do the following or is the cost the same?
- NFP sets status in DB for create/delete done
- node driver polls on status in DB